tls/s2n_kem.c

```diff
@@ -496,3 +442,23 @@ int s2n_kem_recv_ciphertext(struct s2n_stuffer *in, struct s2n_kem_params *kem_p
 
     return S2N_SUCCESS;
 }
+
+S2N_RESULT s2n_kem_init()
+{
+    for (size_t i = 0; i < s2n_array_len(ALL_SUPPORTED_KEM_GROUPS); i++) {
+        struct s2n_kem_group *group = ALL_SUPPORTED_KEM_GROUPS[i];
+        group->available = s2n_pq_is_enabled();
+        /* Only Kyber768+ requires s2n_libcrypto_supports_kyber() */
+        /* TODO: remove the conditional guard when we remove the interned Kyber512 impl. */
+        if (strcmp(group->kem->name, "kyber512r3") != 0) {
+            group->available &= s2n_libcrypto_supports_kyber();
+        }
+        /* x25519 based tls13_kem_groups require EVP_APIS_SUPPORTED */
+        if (group->curve == NULL || group->curve->name == NULL) {
+            group->available = false;
+        } else if (strcmp(group->curve->name, "x25519") == 0) {
+            group->available &= s2n_is_evp_apis_supported();
+        }
+    }
+    return S2N_RESULT_OK;
```

Comment:
I wonder if "available" should be a function pointer instead of a bool. That way you can implement it differently for different groups, and you won't need to make your groups non-const everywhere.

Reply:
it's a fair point, and something that should be reconsidered if the EVP-related ECC APIs ever get migrated to runtime detection support. for the purposes of this PR, i think it would be overly complex.

Reply:
I think it might actually simplify your PR. You won't need to switch all your kem groups to non-const, which will really cut down on the changes. We do it [for ciphers](https://github.com/aws/s2n-tls/blob/e9d48da05492e6294fc15a90c7b409b19c98714c/crypto/s2n_cipher.h#L85).

But obviously not a blocker-- what you have matches ciphers suites.

Reply:
Hm, I see what you're saying. We could remove all the initialization logic. i'll give it a shot.

Reply:
I ended up implementing this as a standalone function rather than as a function pointer. We only need one implementation for all the KEM groups.